dataclients/kubernetes/clusterclient.go

```diff
@@ -0,0 +1,374 @@
+	if err != nil {
+		return nil, err
+	}
+
+	c := &clusterClient{
+		ingressesURI:   ingressesClusterURI,
+		routeGroupsURI: routeGroupsClusterURI,
+		servicesURI:    servicesClusterURI,
+		endpointsURI:   endpointsClusterURI,
+		ingressClass:   ingClsRx,
+		httpClient:     httpClient,
+		token:          token,
+		apiURL:         apiURL,
+	}
+
+	if o.KubernetesNamespace != "" {
+		c.setNamespace(o.KubernetesNamespace)
+	}
+
+	return c, nil
+}
+
+func (c *clusterClient) setNamespace(namespace string) {
+	c.ingressesURI = fmt.Sprintf(ingressesNamespaceFmt, namespace)
+	c.routeGroupsURI = fmt.Sprintf(routeGroupsNamespaceFmt, namespace)
+	c.servicesURI = fmt.Sprintf(servicesNamespaceFmt, namespace)
+	c.endpointsURI = fmt.Sprintf(endpointsNamespaceFmt, namespace)
+}
+
+func (c *clusterClient) createRequest(uri string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest("GET", c.apiURL+uri, body)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.token != "" {
+		req.Header.Set("Authorization", "Bearer "+c.token)
+	}
+
+	return req, nil
+}
+
+func (c *clusterClient) getJSON(uri string, a interface{}) error {
+	log.Debugf("making request to: %s", uri)
+
+	req, err := c.createRequest(uri, nil)
+	if err != nil {
+		return err
+	}
+
+	rsp, err := c.httpClient.Do(req)
+	if err != nil {
+		log.Debugf("request to %s failed: %v", uri, err)
+		return err
+	}
+
+	log.Debugf("request to %s succeeded", uri)
+	defer rsp.Body.Close()
+
+	if rsp.StatusCode == http.StatusNotFound {
+		return errResourceNotFound
+	}
+
+	if rsp.StatusCode != http.StatusOK {
+		log.Debugf("request failed, status: %d, %s", rsp.StatusCode, rsp.Status)
+		return fmt.Errorf("request failed, status: %d, %s", rsp.StatusCode, rsp.Status)
+	}
+
+	b := bytes.NewBuffer(nil)
+	if _, err = io.Copy(b, rsp.Body); err != nil {
+		log.Debugf("reading response body failed: %v", err)
+		return err
+	}
+
+	err = json.Unmarshal(b.Bytes(), a)
+	if err != nil {
+		log.Debugf("invalid response format: %v", err)
+	}
+
+	return err
+}
+
+func (c *clusterClient) clusterHasRouteGroups() (bool, error) {
+	var crl clusterResourceList
+	if err := c.getJSON(clusterZalandoResourcesURI, &crl); err != nil {
+		return false, err
+	}
+
+	for _, cr := range crl.Items {
+		if cr.Name == routeGroupsName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// filterIngressesByClass will filter only the ingresses that have the valid class, these are
+// the defined one, empty string class or not class at all
+func (c *clusterClient) filterIngressesByClass(items []*ingressItem) []*ingressItem {
+	validIngs := []*ingressItem{}
+
+	for _, ing := range items {
+		// No metadata is the same as no annotations for us
+		if ing.Metadata != nil {
+			cls, ok := ing.Metadata.Annotations[ingressClassKey]
+			// Skip loop iteration if not valid ingress (non defined, empty or non defined one)
+			if ok && cls != "" && !c.ingressClass.MatchString(cls) {
+				continue
+			}
+		}
+		validIngs = append(validIngs, ing)
```

Comment:
So if we have no metadata we add it anyways (if we do not delete the metadata check, then we have to put this into the if branch.

Reply:
well, I think that for filtering on ingress classes, this is actually fine. I would like to have though full validation of the ingress objects, in a PR specifically for this. This code here is just being moved from kube.go with the current change. (https://github.com/zalando/skipper/blob/master/dataclients/kubernetes/kube.go#L1246)

Reply:
ok